### PR TITLE
Buy&Sell methods for Uniques

### PR DESCRIPTION
**Pull request by jsidorenko**
This PR adds new buy&sell methods for NFTs.

Cumulus Companion: https://github.com/paritytech/cumulus/pull/1324

### CODE REVIEW

File: frame/uniques/src/lib.rs [hunk 1]
```diff
@@ -1338,5 +1406,51 @@ pub mod pallet {
...
+		#[pallet::weight(0)]
+		pub fn buy_item(
+			origin: OriginFor<T>,
+			class: T::ClassId,
+			instance: T::InstanceId,
+			bid_price: BalanceOrAssetOf<T, I>,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			Self::do_buy_item(class, instance, origin, bid_price)?;
+			Ok(())
```
Comment (ggwpez) [inline frame/uniques/src/lib.rs:1453]:
```suggestion
			Self::do_buy_item(class, instance, origin, bid_price)
```

File: frame/uniques/src/types.rs [hunk 2]
```diff
@@ -125,3 +131,40 @@ pub struct InstanceMetadata<DepositBalance, StringLimit: Get<u32>> {
...
+	A: PartialEq,
+	B: PartialOrd,
+	AB: PartialOrd,
+{
+	pub fn is_greater_or_equal(&self, other: &Self) -> bool {
+		use BalanceOrAsset::*;
+		match (self, other) {
+			(Balance { amount: a }, Balance { amount: b }) => a >= b,
+			(Asset { amount: a, .. }, Asset { amount: b, .. }) => a >= b,
+			_ => false,
```
Comment (ggwpez) [inline frame/uniques/src/types.rs:159]: You could also return an `WrongCurrency` here and throw out the `is_same_currency`.

File: frame/uniques/src/functions.rs [hunk 3]
```diff
@@ -183,8 +188,94 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+		ensure!(details.owner != buyer, Error::<T, I>::NoPermission);
+
+		let price_info =
+			InstancePriceOf::<T, I>::get(&class, &instance).ok_or(Error::<T, I>::NotForSale)?;
+
+		ensure!(bid_price.is_same_currency(&price_info.0), Error::<T, I>::WrongCurrency);
+		ensure!(bid_price.is_greater_or_equal(&price_info.0), Error::<T, I>::ItemUnderpriced);
+
+		if let Some(only_buyer) = price_info.1 {
+			ensure!(only_buyer == buyer, Error::<T, I>::NotForSale);
```
Comment (ggwpez) [inline frame/uniques/src/functions.rs:239]:
```suggestion
			ensure!(only_buyer == buyer, Error::<T, I>::NoPermission);
```

File: frame/uniques/src/lib.rs [hunk 4]
```diff
@@ -1338,5 +1406,51 @@ pub mod pallet {
...
+		///
+		/// Origin must be Signed and must not be the owner of the asset `instance`.
+		///
+		/// - `class`: The class of the asset.
+		/// - `instance`: The instance of the asset the sender wants to buy.
+		/// - `bid_price`: The price the sender is willing to pay.
+		///
+		/// Emits `ItemBought` on success.
+		#[pallet::weight(0)]
+		pub fn buy_item(
```
Comment (ggwpez) [inline frame/uniques/src/lib.rs:1445]:
```suggestion
		#[transactional]
		pub fn buy_item(
```

File: frame/uniques/src/lib.rs [hunk 5]
```diff
@@ -329,6 +369,26 @@ pub mod pallet {
 		},
 		/// Ownership acceptance has changed for an account.
 		OwnershipAcceptanceChanged { who: T::AccountId, maybe_class: Option<T::ClassId> },
+		/// The price was set for the instance.
+		InstancePriceSet {
+			class: T::ClassId,
+			instance: T::InstanceId,
+			price: BalanceOrAssetOf<T, I>,
+			buyer: Option<T::AccountId>,
```
Comment (ggwpez) [inline frame/uniques/src/lib.rs:377]:
```suggestion
			whitelisted_buyer: Option<T::AccountId>,
```
? since he did not buy it *yet*

**Review by ggwpez — COMMENTED** [review]
I did not look at the tests yet, but everything else looks great!

File: frame/uniques/src/lib.rs [hunk 6]
```diff
@@ -150,6 +153,31 @@ pub mod pallet {
 
 		/// Weight information for extrinsics in this pallet.
 		type WeightInfo: WeightInfo;
+
+		/// Fungible assets.
+		type Assets: Inspect<Self::AccountId, AssetId = Self::AssetId, Balance = Self::AssetBalance>
+			+ Transfer<Self::AccountId>;
+
+		/// Fungible asset balance.
+		type AssetBalance: sp_runtime::traits::AtLeast32BitUnsigned
```
[the remark this answers is not in the document]
Comment (xlc) [reply frame/uniques/src/lib.rs:162]: https://github.com/paritytech/substrate/blob/76a6cc3ca66926fa88569c2b5bfe2b311ef335ae/frame/support/src/traits/tokens/misc.rs#L168-L169

**Review by xlc — COMMENTED** [review]
it shouldn't be the uniques pallet handling the difference between balances and assets. this simply doesn't scale.

I will suggest to take a look or how orml handles it. http://github.com/xlc/orml-workshop

Basically the pallet will only deal with a multi asset trait (`fungibles::*`) and that's it.

We have another struct / pallet that deal with detecting which pallet by looking at the asset id and relay to the right pallet.

**Comment by gilescope** [comment]
Does this functionality overlap with the atomic-swap pallet? Maybe improvements to that pallet might be an option if it doesn't do all you would wish it to?

File: frame/uniques/src/types.rs (hunk 2 above)
Comment (jsidorenko) [reply frame/uniques/src/types.rs]: Done

File: frame/uniques/src/functions.rs [hunk 7]
```diff
@@ -195,8 +201,85 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+		ensure!(!details.is_frozen, Error::<T, I>::Frozen);
+		ensure!(details.owner == sender, Error::<T, I>::NoPermission);
+
+		if let Some(ref price) = price {
+			ItemPriceOf::<T, I>::insert(
+				&collection,
+				&item,
+				(price.clone(), whitelisted_buyer.clone()),
+			);
+			Self::deposit_event(Event::ItemPriceSet {
```
Comment (ggwpez) [inline frame/uniques/src/functions.rs:232]: I think we should only emit an even if the price changed.  
This would incur an additional storage read though.

File: frame/uniques/src/functions.rs [hunk 8]
```diff
@@ -195,8 +201,85 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+		ensure!(bid_price >= price_info.0, Error::<T, I>::BidTooLow);
+
+		if let Some(only_buyer) = price_info.1 {
+			ensure!(only_buyer == buyer, Error::<T, I>::NoPermission);
+		}
+
+		T::Currency::transfer(
+			&buyer,
+			&details.owner,
+			bid_price.clone(),
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply frame/uniques/src/functions.rs:267]: Yes, I agree. The transfer price here should be set price on the NFT, not the overbid by the user.

File: frame/uniques/src/functions.rs (hunk 7 above)
Comment (shawntabrizi) [reply frame/uniques/src/functions.rs:232]: This isnt worth it imo. I believe that the difference between just writing here vs read then write will increase the PoV size and stuff like that. Chances are, user is not uselessly setting the price to be the same, and if they do, we don't care.

File: frame/uniques/src/functions.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply frame/uniques/src/functions.rs]: You fixed it thanks, is there a test that checks the balance after this buy call? Otherwise there probably should be.

[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/uniques/src/functions.rs]: Done, modified the existing test to cover that edge-case

File: frame/uniques/src/functions.rs [hunk 9]
```diff
@@ -196,8 +202,85 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+		price: Option<ItemPrice<T, I>>,
+		whitelisted_buyer: Option<T::AccountId>,
+	) -> DispatchResult {
+		let collection_details =
+			Collection::<T, I>::get(&collection).ok_or(Error::<T, I>::UnknownCollection)?;
+		ensure!(!collection_details.is_frozen, Error::<T, I>::Frozen);
+		ensure!(!T::Locker::is_locked(collection, item), Error::<T, I>::Locked);
+
+		let details = Item::<T, I>::get(&collection, &item).ok_or(Error::<T, I>::UnknownItem)?;
+		ensure!(!details.is_frozen, Error::<T, I>::Frozen);
```
Comment (shawntabrizi) [inline frame/uniques/src/functions.rs:224]: I don't think you should check frozen or locked for updating the price. Users should be able to update the price, even if the asset if frozen, thus when it unfreezes, it can be traded at the appropriate price.

**Comment by joao-paulo-parity** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_uniques

**Review by shawntabrizi — APPROVED** [review]
